judge/views/problem.py

```diff
@@ -201,9 +201,69 @@ def get_context_data(self, **kwargs):
                                           context['description'], 'problem')
         context['meta_description'] = self.object.summary or metadata[0]
         context['og_image'] = self.object.og_image or metadata[1]
+
+        context['can_vote'] = self.object.can_vote(user)
+
+        # Get the vote this user has already cast on this problem.
+        if context['can_vote']:
+            try:
+                context['vote'] = ProblemPointsVote.objects.get(voter=user.profile, problem=self.object)
+            except ObjectDoesNotExist:
+                context['vote'] = None
+        else:
+            context['vote'] = None
+
+        all_votes = list(self.object.problem_points_votes.order_by('points').values_list('points', flat=True))
+        context['has_votes'] = len(all_votes) > 0
+
+        if not self.request.in_contest:
+            context['all_votes'] = all_votes
+
+        context['max_possible_vote'] = settings.DMOJ_PROBLEM_MAX_USER_POINTS_VOTE
+        context['min_possible_vote'] = max(settings.DMOJ_PROBLEM_MIN_USER_POINTS_VOTE,
+                                           settings.DMOJ_PROBLEM_MIN_PROBLEM_POINTS)
         return context
 
 
+class DeleteProblemVote(ProblemMixin, SingleObjectMixin, View):
+    def get(self, request, *args, **kwargs):
+        return HttpResponseForbidden(status=405, content_type='text/plain')
+
+    def post(self, request, *args, **kwargs):
+        self.object = self.get_object()
+        if not request.user.is_authenticated:
+            return HttpResponseForbidden(_('Not signed in.'), content_type='text/plain')
+        elif self.object.can_vote(request.user):
+            ProblemPointsVote.objects.filter(voter=request.profile, problem=self.object).delete()
+            return HttpResponse('success', content_type='text/plain')
+
+        return HttpResponseForbidden(_('Not allowed to delete votes on this problem.'), content_type='text/plain')
+
+
+class ProblemVote(ProblemMixin, SingleObjectMixin, View):
+    def get(self, request, *args, **kwargs):
+        return HttpResponseForbidden(status=405, content_type='text/plain')
+
+    def post(self, request, *args, **kwargs):
+        self.object = self.get_object()
+        if not self.object.can_vote(request.user):
```

Comment:
For deleting votes, we have a separate `if` checking if the user is authenticated while we don't do that here. We should either add that `if` in both places or remove it from from `DeleteProblemVote` and let the user authentication check be handled by `can_vote`.

Reply:
Let's do the latter.

Reply:
I removed the check from DeleteProblemVote.